python/python/bystro/prs/preprocess_for_prs.py

```diff
@@ -305,48 +281,60 @@ def ld_clump(scores_overlap: pd.DataFrame, map_directory_path: str) -> tuple[pd.
     return clean_scores_for_analysis(max_effect_per_bin, "ID_effect_as_ref")
 
 
-def extract_clumped_thresholded_genos(
-    dosage_matrix_path: str, scores_after_c_t: pd.DataFrame
-) -> pd.DataFrame:
-    dosage_dataset = ds.dataset(dosage_matrix_path, format="feather")
-    loci_list = scores_after_c_t.index.tolist()
-    loci_filter = ds.field("locus").isin(loci_list)
-    filtered_dataset = dosage_dataset.filter(loci_filter)
-    filtered_df = filtered_dataset.to_table().to_pandas()
-    return filtered_df
-
-
-def finalize_dosage_scores_after_c_t(
+def finalize_scores_after_c_t(
     gwas_scores_path: str, dosage_matrix_path: str, map_directory_path: str, p_value_threshold: float
 ) -> tuple[pd.DataFrame, pd.DataFrame]:
-    """Finalize dosage matrix and scores for PRS calculation."""
-    scores_overlap = generate_thresholded_overlap_scores_dosage(
-        gwas_scores_path, dosage_matrix_path, p_value_threshold
-    )
+    """Finalize scores for PRS calculation."""
+    scores = _load_association_scores(gwas_scores_path)
+    preprocessed_scores = _preprocess_scores(scores)
+    thresholded_score_loci = get_p_value_thresholded_indices(preprocessed_scores, p_value_threshold)
+    dosage_loci_nomiss = _extract_nomiss_dosage_loci(dosage_matrix_path, thresholded_score_loci)
+    overlap_loci = generate_overlap_scores_dosage(thresholded_score_loci, dosage_loci_nomiss)
+    scores_overlap = preprocessed_scores[preprocessed_scores.index.isin(overlap_loci)]
     scores_after_c_t, loci_and_allele_comparison = ld_clump(scores_overlap, map_directory_path)
-    dosage_overlap = extract_clumped_thresholded_genos(dosage_matrix_path, scores_after_c_t)
-    dosage_overlap = dosage_overlap.set_index("locus")
+    scores_after_c_t = scores_after_c_t.sort_index()
+    return scores_after_c_t, loci_and_allele_comparison
+
+
+def finalize_dosage_after_c_t(
+    dosage_overlap: pd.DataFrame, loci_and_allele_comparison: pd.DataFrame
+) -> pd.DataFrame:
+    """Finalize dosage matrix for PRS calculation."""
     merged_allele_comparison_genos = loci_and_allele_comparison.join(dosage_overlap, how="inner")
     merged_allele_comparison_genos = merged_allele_comparison_genos.sort_index()
-    scores_after_c_t = scores_after_c_t.sort_index()
     genos_adjusted = merged_allele_comparison_genos.apply(adjust_dosages, axis=1)
     genotypes_adjusted_only = genos_adjusted.iloc[:, 1:]
     genos_transpose = genotypes_adjusted_only.T
-    return genos_transpose, scores_after_c_t
+    return genos_transpose
 
 
 def generate_c_and_t_prs_scores(
```

Comment:
Let's simplify this a bit; we don't need the read_feather_in_chunks anymore, and can directly filter in pyarrow which will be faster. Please try this and update your test:

```suggestion
def generate_c_and_t_prs_scores(
    gwas_scores_path: str,
    dosage_matrix_path: str,
    map_directory_path: str,
    p_value_threshold: float = 0.05,
) -> dict[str, float]:
    """Calculate PRS."""
    # This part goes through dosage matrix the first time to get overlapping loci
    scores_after_c_t, loci_and_allele_comparison = finalize_scores_after_c_t(
        gwas_scores_path, dosage_matrix_path, map_directory_path, p_value_threshold
    )

    # This part goes through dosage matrix the second time, adjusts dosages,
    # transposes genotypes, and calculates PRS
    prs_scores: pd.Series = pd.Series(dtype=float)
    beta_values = scores_after_c_t["BETA"]
    loci = scores_after_c_t.index

    score_loci_filter = pc.field("locus").isin(pa.array(list(loci)))
    dosage_ds = ds.dataset(dosage_matrix_path, format="feather").filter(score_loci_filter)

    for batch in dosage_ds.to_batches(batch_size=1000, batch_readahead=1):
        chunk = batch.to_pandas()
        if chunk.empty:
            continue

        chunk = chunk.set_index("locus")
        genos_transpose = finalize_dosage_after_c_t(batch, loci_and_allele_comparison)
        prs_scores_chunk = genos_transpose @ beta_values.loc[genos_transpose.columns]
        prs_scores = prs_scores.add(prs_scores_chunk, fill_value=0)

    return prs_scores.to_dict()
```